stdlib/public/core/Hashable.swift

```diff
@@ -185,10 +185,11 @@ public struct _UnsafeHasher {
     return value._hash(into: self)
   }
 
-  @_inlineable
+  @effects(readonly) // See comment on appending(_:) above
+  @inline(never)
```

Comment:
Why is this inline never? Shouldn't this be inlined into any part of the standard library that's behind a resilience barrier, such as the internals of Dictionary?

Reply:
Indeed! This really only demonstrates how much effect a single additional non-inlineable function call during hashing has on our benchmarks. (By accident, it also highlighted the effect of `@effects(readonly)` rather nicely.) The benchmark below seems like a good target to shoot for.

The size of the `_Hasher` struct and its initializer are both baked into binaries at the moment. The entire `_Hasher` should be unversioned. Turns out, doing that by naively slapping `@inline(never)` on `_defaultHashValue` slows benchmarks down by up to 1800%.

Reply:
If we want to stack allocate the state on the stack, that makes it fixed layout. (Assuming non-alloca solutions here, but perhaps @aschwaighofer's improvements to stack-allocating generic/opaque values from last year would help).

When the user has defined their own `hashValue` on one member of a struct, and is deriving Hashable conformance for the whole struct, is the result of that member's hashValue fed into the `_Hasher` instead of the member's data? Can we partition hashing into a main `_Hasher` that's visiting everything, sometimes temporarily spawning a local `_Hasher` for any nested commutative collections before feeding that result back to the main one? If so, the main one can live on TLS and the embedded ones can be opaque/indirect. Then, we only pay the opaque layout cost once per nested commutative structure.

Reply:
`_Hasher` can be stack-allocated as long as it’s initialized, fed and finalized inside a resilient function, right? It can have a fixed layout and fully inlineable methods as long as it’s not `@_versioned`.

But I’m not yet sure if there is a way to effectively pass a `Hashable`‘s `_hash(into:)` to such a function without incurring unacceptable overhead. Indeed, TLS for the root hasher will likely be the solution if we can’t.